src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -0,0 +1,472 @@
+                                    final SAMFileHeader h = context.sideInput(headerPcolView);
+                                    final String key = MarkDuplicatesReadsKey.keyForFragment(h, record);
+                                    final KV<String, Read> kv = KV.of(key, record);
+                                    context.output(kv);
+                                }
+                            });
+            return fragments.apply(makeKeysForFragments)
+                    .apply(GroupByKey.<String, Read>create());
+        }
+    }
+
+    /**
+     * How to assign a score to the read in MarkDuplicates (so that we pick the best one to be the non-duplicate).
+     */
+    //Note: copied from htsjdk.samtools.DuplicateScoringStrategy
+    private static int score(final Read record) {
+        if (record == null) {
+            return 0;
+        } else {
+            return record.getAlignedQuality().stream().mapToInt(Integer::intValue).filter(i -> i >= MIN_BASE_QUAL).sum();
+        }
+    }
+
+    /**
+     * Returns the list of intervals from the given sequence dictionary.
+     */
+    private static List<SimpleInterval> getAllIntervalsForReference(final SAMSequenceDictionary sequenceDictionary) {
+    return GenomeLocSortedSet.createSetFromSequenceDictionary(sequenceDictionary)
+            .stream()
+            .map(SimpleInterval::new)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Struct-like class to store information about the paired reads for mark duplicates.
+     */
+    private static final class PairedEnds extends GenericJson{
+        private Read first, second;
+
+        PairedEnds(final Read first) {
+            this.first = first;
+        }
+
+        public static PairedEnds of(final Read first) {
+            return new PairedEnds(first);
+        }
+
+        public PairedEnds and(final Read second) {
+            if (second != null && GenomicsReadUtils.unclippedCoordinate(this.first) > GenomicsReadUtils.unclippedCoordinate(second)) {
+                this.second = this.first;
+                this.first = second;
+            } else {
+                this.second = second;
+            }
+            return this;
+        }
+
+        public String key(final SAMFileHeader header) {
+            return MarkDuplicatesReadsKey.keyForPairedEnds(header, first, second);
+        }
+
+        public Read first() {
+            return first;
+        }
+
+        public Read second() {
+            return second;
+        }
+
+        public int score() {
+            return MarkDuplicatesDataflow.score(first) + MarkDuplicatesDataflow.score(second);
+        }
+
+        @Override
+        public PairedEnds clone() {
+            return (PairedEnds) super.clone();
+        }
+    }
+
+    /**
+     * Special coder for the PairedEnds class.
+     */
+    private static final class PairedEndsCoder extends AtomicCoder<PairedEnds> {
```

Comment:
I think this is overly complex. You should be able to either register the GenericJsonCoder for your class or use a Delegate coder. Are you really expecting nulls?

    public static final DelegateCoder<PairedEnds, KV<Read, Read>> CODER =
            DelegateCoder.of(
                    KvCoder.of(GenericJsonCoder.of(Read.class), GenericJsonCoder.of(Read.class)),
                    new DelegateCoder.CodingFunction<PairedEnds, KV<Read, Read>>() {
                        @Override
                        public KV<Read, Read> apply(PairedEnds ends) throws Exception {
                            return KV.of(ends.first(), ends.second());
                        }
                    },
                    new DelegateCoder.CodingFunction<KV<Read, Read>, PairedEnds>() {
                        @Override
                        public ReferenceShard apply(KV<Read, Read> kv) throws Exception {
                            return PairedEnds.of(kv.getKey()).and(kv.getValue());
                        }
                    }
            );